Ensure subclasses inherit fields

chrismytton opened the pull request:
This adds an `inherited` hook to ensure that subclasses inherit fields correctly.

Fixes #6

chrismytton committed:
Add failing tests for field inheritance bug

ondenman committed:
Include parent class fields in to_h result

Previously, if a class inherited from a parent class with fields
defined in it, the subclass would not return the parent class fields in
its #to_h result:

InheritanceTest#methods: constituency
InheritanceSubclassTest#methods: state

> to_h
=> {:state=>"Zamfara"}

This commit fixes the bug.

> to_h
=> {:constituency=>"Bungudu", :state=>"Zamfara"}

tmtmtmtm commented on test/field_serializer_test.rb:
```diff
@@ -68,4 +68,24 @@ class InternalFieldTest
+      InheritanceSubclassTest.new.to_h.must_equal(constituency: 'Bungudu', state: 'Zamfara')
+    end
+  end
 end
```
I think you also want a test here to show that it DTRT in cases where a field exists in both the parent and the subclass…

ondenman committed:
Test the object returns field set in subclass

Tests that field set in subclass overrides that of the same name set in parent class.

ondenman commented:
@tmtmtmtm Added test to show that field in subclass overrides field in parent class.

ondenman committed:
Rename tests for clarity

chrismytton commented:
@ondenman tl;dr I think this can probably be done in a simpler way using Ruby's meta-programming tools. I've written in a bit more detail below to hopefully give you some insight into which tools in particular might be suitable for this job and why.

---

As you may know, Ruby has "class variables", "instance variables" and "class instance variables". The current implementation of `FieldSerializer` (on the `master` branch) uses "class instance variables". This is why fields aren't currently being inherited by subclasses.

My initial thought was that this _might_ be fixable by simply switching to "class variables".

``` diff
diff --git a/lib/field_serializer.rb b/lib/field_serializer.rb
index 2f493dd..98e9af8 100644
--- a/lib/field_serializer.rb
+++ b/lib/field_serializer.rb
@@ -12,7 +12,7 @@ module FieldSerializer
   module ClassMethods

     def fields
-      @fields ||= []
+      @@fields ||= []
     end

     def field(name, &block)
```

This doesn't work! 😨  

It turns out (and this was new to me!) that if you use class variables in a module such as `FieldSerializer`, then the `@@fields` class variable is defined _on the `FieldSerializer` module_, not on the class the module is mixed into.

We can fix this problem with Ruby's metaprogramming tools. Using `class_variable_defined?`, `class_variable_set` and `class_variable_get` <sup>[1](#fn-1)</sup> allows us to set and get "class variables" on the class that `FieldSerializer` is mixed into, rather than on the module.

``` diff
diff --git a/lib/field_serializer.rb b/lib/field_serializer.rb
index 2f493dd..caac978 100644
--- a/lib/field_serializer.rb
+++ b/lib/field_serializer.rb
@@ -12,7 +12,11 @@ module FieldSerializer
   module ClassMethods

     def fields
-      @fields ||= []
+      if class_variable_defined?(:@@fields)
+        class_variable_get(:@@fields)
+      else
+        class_variable_set(:@@fields, [])
+      end
     end

     def field(name, &block)
```

This works because `class_variable_defined?` and friends are being called on `self`, which at that point will be the target class.

---

<a name="fn-1">1</a>: [This blog post explains these in a bit more detail.](http://thoughts.codegram.com/understanding-class-instance-variables-in-ruby/)

tmtmtmtm commented:
@chrismytton I think this approach will crash horribly if something then further subclasses a subclass, e.g.

``` patch
diff --git a/test/field_serializer_test.rb b/test/field_serializer_test.rb
index 1cf43b5..ba082c7 100644
--- a/test/field_serializer_test.rb
+++ b/test/field_serializer_test.rb
@@ -94,6 +94,20 @@ describe FieldSerializer  do
       end
     end

+    class InheritanceSubclassTest3 < InheritanceSubclassTest2
+      field :constituency do
+        'Wibble'
+      end
+
+      field :district do
+        'Bar'
+      end
+    end
+
+    it 'should allow subclasses to be subclassed' do
+      InheritanceSubclassTest3.new.to_h.must_equal(constituency: 'Wibble', state: 'Zamfara', district: 'Bar')
+    end
+
     it 'should return fields defined by parent class in #to_h' do
       InheritanceSubclassTest.new.to_h.must_equal(constituency: 'Bungudu', state: 'Zamfara')
```

Because of the way class variables are shared, now the parent classes of `InheritanceSubclassTest3` all believe they should have a `:district` field, so `InheritanceSubclassTest2` blows up with a `undefined method` error.

chrismytton commented:
@tmtmtmtm 😱  You're quite right! Hmm, in that case I'm not sure what the best solution is here. It certainly seems like something that _should_ be possible without having to jump through hoops calling `super`, but nothing is jumping to mind immediately.

chrismytton committed:
Add inheritence test for subclasses of subclasses

This tests that subclasses of subclasses don't introduce fields into the
parent class.

chrismytton committed:
Add FieldSerializer::ClassMethods#inherited hook

This allows us to run some code when a subclass is created, in this case
we load up the subclass with the `fields` attribute from the parent
class.

I was guided to this solution by this answer on StackOverflow:
http://stackoverflow.com/a/13629724

chrismytton commented:
I've pushed a couple of commits which make this change a bit simpler by using Ruby's `inherited` hook. I've also added the test that @tmtmtmtm suggested.

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm committed:
Merge branch 'master' into subclasses-inherit-fields

tmtmtmtm commented:
Woot! Much much nicer now! I was going to say that this probably needs an entry in the Changelog, but it seems there isn't one. 

We probably should update the README though, as this is a fairly significant feature. (Though I'm OK with this being merged now and the README added separately if this is blocking something)
